src/wal/mod.rs

```diff
@@ -129,3 +123,136 @@ impl Wal {
         Ok(())
     }
 }
+
+/// Open a WAL file and check the integrity of the file, it decodes
+/// all records and make sure they are coherent with the WAL format.
+pub struct WalChecker {
+    last_batch: Option<Batch>,
+}
+
+pub enum ConsistencyError {
+    LastBatchCrashed(Batch),
+    NotConsistent(u64),
+}
+
+impl WalChecker {
+    /// Open a WAL file and check its integrity, if a problem is found then the
+    /// wal is restored to its last valid state, thus it updates the WAL file
+    /// by removing everything after the last valid record.
+    ///
+    /// NOTE: Even though currently in the simulation each batch is pruned right after
+    /// the success of the next one, the following method needs to be able to handle multiple
+    /// records of different batches and validate them
+    pub fn open_and_recorver(path: PathBuf) -> Self {
+        let mut wal_file = match OpenOptions::new().read(true).write(true).open(&path) {
+            Ok(wal_file) => wal_file,
+            Err(err) if err.kind() == std::io::ErrorKind::NotFound => {
+                // a non-existing WAL is considered valid, with the assumed sequence number being zero
+                return Self { last_batch: None };
+            }
+            Err(_) => {
+                panic!("Error opening or creating the wal file ")
+            }
+        };
+
+        let (last_records, last_records_position) = Self::validate_records(&mut wal_file);
+
+        // If not all the WAL was parsed correctly, then the recovery procedure means to remove
+        // all incorrect records from the last valid one
+
+        let wal_file_len = wal_file
+            .metadata()
+            .expect("Error extracting wal file len")
+            .len();
+
+        if last_records_position as u64 != wal_file_len {
+            // cannot use fallocate + FALLOC_FL_COLLAPSE_RANGE
+            // because it requires offset and len to be multiple of the
+            // filesystem block size, and len could be different from a multiple of a record size
+            unsafe {
+                let res =
+                    libc::ftruncate(wal_file.as_raw_fd(), last_records_position as libc::off_t);
+                assert!(res == 0, "WAL Recovery: ftruncate failed");
+            }
+
+            wal_file.sync_all().unwrap();
+        }
+
+        Self {
+            last_batch: Some(Batch::from_records(last_records)),
+        }
+    }
+
+    fn validate_records(wal_file: &mut File) -> (Vec<Record>, usize) {
```

Comment:
style: for the future - I've never much liked the approach of putting functions on a struct which don't construct the type itself. This could just as easily be a free function.

Reply:
Yeah, it makes sense! It is definitely Java-like

For me, the reason is that it is something that is not really reused somewhere else, and its only reason to exist is that it makes the `open_and_recover` function more readable. Thus, associating it with the `WalChecker` makes the structure of the code clearer

But, I mean, it's just a function. It will be moved